File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
def solution(N, number):
    
    dp = [set() for _ in range(8)]
    
    for i in range(len(dp)):
        dp[i].add(int(str(N) * (i + 1)))
    
    for i in range(len(dp)):
        for j in range(i):
            for a in dp[j]:
                for b in dp[i - j - 1]:
                    dp[i].add(a + b)
                    dp[i].add(a - b)
                    dp[i].add(b - a)
                    dp[i].add(a * b)
                    
                    if a != 0:
                        dp[i].add(b // a)
                    if b != 0:
                        dp[i].add(a // b)
                        
            if j == i - j - 1:
                break
        
        if number in dp[i]:
            return i + 1

    
    return -1

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
import java.util.*;

class Solution {
    public int solution(int N, int number) {
        
        HashSet<Integer>[] dp = new HashSet[8];
        
        for (int i = 0; i < dp.length; i++) {
            dp[i] = new HashSet<>();
            dp[i].add(Integer.parseInt(String.valueOf(N).repeat(i + 1)));
        }

        for (int i = 0; i < dp.length; i++) {
            
            for (int j = 0; j < i; j++) {

                for (int a : dp[j]) {
                    for (int b : dp[i - j - 1]) {

                        dp[i].add(a + b);
                        dp[i].add(a - b);
                        dp[i].add(b - a);
                        dp[i].add(b * a);

                        if (a != 0) {
                            dp[i].add(b / a);
                        }
                        if (b != 0) {
                            dp[i].add(a / b);
                        }
                    }                                                        
                }
            }
            
            if (dp[i].contains(number)) {
                 return i + 1;
            }
            
        }

        return -1;
    }
}